include/rmm/mr/device/detail/stream_ordered_memory_resource.hpp

```diff
@@ -356,29 +362,46 @@ class stream_ordered_memory_resource : public crtp<PoolResource>, public device_
    * @return A block with non-null pointer and size >= `size`, or a nullptr block if none is
    *         available in `blocks`.
    */
-  block_type get_block_from_other_stream(size_t size, stream_event_pair stream_event)
+  block_type get_block_from_other_stream(size_t size,
+                                         stream_event_pair stream_event,
+                                         free_list& blocks,
+                                         bool merge_first)
   {
-    // nothing in this stream's free list, look for one on another stream
-    for (auto s = stream_free_blocks_.begin(); s != stream_free_blocks_.end(); ++s) {
-      auto blocks_event = s->first;
-      if (blocks_event.event != stream_event.event) {
-        auto blocks = s->second;
-
-        block_type const b = blocks.get_block(size);  // get the best fit block
+    for (auto it = stream_free_blocks_.begin(), next_it = it; it != stream_free_blocks_.end();
```

Comment:
Would be nice if we had an `adjacent_for_each` (just a comment - not a request)

Reply:
I don't think it would matter. The whole point of the structure of this loop is to allow deleting an iterator.

Reply:
Oh that is even better - `adjacent_remove_if` would be the algorithm we would want. Range-v3 has this exact algorithm: 
>views::adjacent_remove_if
For each pair of adjacent elements in a source range, evaluate the specified binary predicate. If the predicate evaluates to true, the first element of the pair is removed from the result range; otherwise, it is included. The last element in the source range is always included.
https://ericniebler.github.io/range-v3/index.html#tutorial-views

Reply:
No, because we want to do more on each list than just remove it. 

 `<algorithm>` is your hammer, @codereport, but the universe is more than just nails. :)